tests/unit/error/merr_test.c

```diff
@@ -118,9 +118,12 @@ MTF_DEFINE_UTEST(hse_err_test, no_buf)
     size_t needed_sz;
 
     err = merr(EINVAL);
-
     merr_strinfo(err, NULL, 0, &needed_sz);
     ASSERT_EQ(55, needed_sz);
+
+    err = merr(EBUG);
+    merr_strinfo(err, NULL, 0, &needed_sz);
+    ASSERT_EQ(56, needed_sz);
```

Comment:
i still think that needed-size argument is unnecessary and just adds unnecessary complexity....
You prob should have a comment here saying these magic values are the string lengths of....

Reply:
hmm, yeah it would be more maintainable to snprintf the expected string to get the size you expect. Let me do that instead of being lazy.

> i still think that needed-size argument is unnecessary and just adds unnecessary complexity

I don't understand how someone is supposed to know if their string is truncated

Reply:
Most use cases are just going to pass in a generously sized buffer and never check for truncation (like strerror).   I just don't think all the extra LOC and complexity pays for itself, it feels to me this is one of those "what more can we add?" scenarios, whereas I'm more in the "when nothing can be taken away" camp (ala Antoine de Saint-Exupéry).

Reply:
Also worth noting that these are internal APIs, so they need not be designed for public consumption.

Reply:
merr_strinfo() is consumed by an external API. `hse_strerror()`. How else do you get the needed size if it doesn't come from this function?

Reply:
Same way you get the needed size from strerror_r() :-)
IMO hse_strerror() API is broken, it should mimic strerror_r()...

Reply:
Assuming you are talking about GNU strerror_r and not the XSI one, then hse_strerror() has completely different constraints.

> The GNU-specific strerror_r() returns a pointer to a string containing the error message. This may be either a pointer to a string that the function stores in buf, or a pointer to some (immutable) static string (in which case buf is unused). If the function stores a string  in  buf,  then at most buflen bytes are stored (the string may be truncated if buflen is too small and errnum is unknown). The string always includes a terminating null byte ('\0').

We could never have a static string to point to, so returning a `char *` makes no sense to me.

Reply:
Typically one just returns the buffer pointer given to the function so that they can use the call inline to some other function...

Reply:
I don't understand why someone wouldn't just use buf. What are you really gaining by returning a `char *`? The ability to call it within a printf() function? Seems like it would just hurt readability to me.

Reply:
I think one or two lines along the lines of

```
    fprintf(stderr, "put 1 failed: %s\n", hse_strerror(err, buf, sizeof(buf)));
```

is IMO vastly more readable than having 5 or more additional lines for every error message I want to print,
which gets worse the more I call it...

```
    n = hse_strerror(err, buf, sizeof(buf), &bufsz);
    if (n >= sizeof(buf)) {
        /* maybe do something, dunno, .. */
        return;
    }

    fprintf(stderr, "put 1 failed: %s\n", buf);
```

Reply:
I would code it so that it doesn't fail, or at least never returns NULL.
Look at strcpy, strcat, whatever, they don't fail if you give them good inputs, and neither do they check their inputs (I think they use __nonnull in the definition).
I strive to create simple to use functions that endeavor not to fail whenever possible, this just seems like one of those cases.  But I cannot convice you, so please ignore me :-)

Reply:
If you don't consider buffer truncation a failure, you can ignore the return value. In that sense, it is infallible like you want. No one is forcing the caller to check for truncation. We only give them that ability.

> Look at strcpy, strcat, whatever, they don't fail if you give them good inputs, and neither do they check their inputs (I think they use __nonnull in the definition).

The simplicity of these APIs is exactly why they are security nightmares and should have been removed decades ago in my opinion.